### PR TITLE
add support for job outputs.

### PR DESCRIPTION
**Pull request by TingluoHuang**
https://github.com/github/c2c-actions-runtime/pull/331

### CODE REVIEW

File: src/Sdk/DTPipelines/workflow-v1.0.json [hunk 1]
```diff
@@ -1,658 +1,762 @@
...
-
-    "workflow-root": {
-      "description": "Workflow file",
-      "mapping": {
-        "properties": {
-          "on": "any",
-          "name": "string",
-          "env": "workflow-env",
-          "jobs": "jobs"
+    "version": "workflow-v1.0",
```
Comment (ericsciple) [inline src/Sdk/DTPipelines/workflow-v1.0.json]: did indent level change?

File: src/Runner.Worker/JobExtension.cs [hunk 2]
```diff
@@ -242,6 +243,59 @@ public void FinalizeJob(IExecutionContext jobContext, Pipelines.AgentJobRequestM
                     context.Start();
                     context.Debug("Starting: Complete job");
 
+                    // Evaluate job outputs
+                    if (message.JobOutputs != null)
```
Comment (ericsciple) [inline src/Runner.Worker/JobExtension.cs]: also check .Type != TokenType.Null ?

File: src/Runner.Worker/JobExtension.cs [hunk 3]
```diff
@@ -242,6 +243,59 @@ public void FinalizeJob(IExecutionContext jobContext, Pipelines.AgentJobRequestM
...
+                            context.ExpressionValues["env"] = envContext;
+                            foreach (var pair in context.EnvironmentVariables)
+                            {
+                                envContext[pair.Key] = new StringContextData(pair.Value ?? string.Empty);
+                            }
+
+                            Trace.Info("Initialize steps context for evaluating job outputs");
+                            context.ExpressionValues["steps"] = context.StepsContext.GetScope(context.ScopeName);
+
+                            var templateTrace = context.ToTemplateTraceWriter();
```
Comment (ericsciple) [inline src/Runner.Worker/JobExtension.cs]: it looks like you dont use this local var

File: src/Sdk/DTPipelines/workflow-v1.0.json (hunk 1 above)
Comment (TingluoHuang) [reply src/Sdk/DTPipelines/workflow-v1.0.json]: vscode did this, I will undo it.

File: src/Runner.Worker/JobExtension.cs (hunk 3 above)
Comment (TingluoHuang) [reply src/Runner.Worker/JobExtension.cs]: forgot to remove this after rebase on master :)

File: src/Runner.Worker/action_yaml.json [hunk 4]
```diff
@@ -88,12 +88,8 @@
         "input-default-context": {
             "context": [
                 "github",
-                "strategy",
```
Comment (ericsciple) [inline src/Runner.Worker/action_yaml.json]: did you mean to do this?

Comment (TingluoHuang) [inline src/Runner.Worker/action_yaml.json]: these don't make sense for input default

Comment (ericsciple) [reply src/Runner.Worker/action_yaml.json]: strategy and matrix are the ones i'm mainly concerned about removing

File: src/Sdk/DTObjectTemplating/ObjectTemplating/Schema/MappingDefinition.cs [hunk 5]
```diff
@@ -30,8 +30,7 @@ internal MappingDefinition(MappingToken definition)
                                     foreach (var propertiesPair in properties)
```
Comment (TingluoHuang) [inline src/Sdk/DTObjectTemplating/ObjectTemplating/Schema/MappingDefinition.cs]: Copied from SDK, not related for outputs but required for support 
```json
foo: 
{
   "requires": "true",
   "type": "runs-on",
   
}
```
in workflow schema.

File: src/Sdk/DTObjectTemplating/ObjectTemplating/Schema/PropertyValue.cs [hunk 6]
```diff
@@ -1,18 +1,40 @@
 ﻿using System;
```
Comment (TingluoHuang) [inline src/Sdk/DTObjectTemplating/ObjectTemplating/Schema/PropertyValue.cs]: Copied from SDK

**Comment by ericsciple** [comment]
looks good. i'll review the sdk stuff on the server pr

File: src/Sdk/DTObjectTemplating/ObjectTemplating/Schema/TemplateSchema.cs [hunk 7]
```diff
@@ -312,8 +312,8 @@ private static TemplateSchema Schema
```
Comment (TingluoHuang) [inline src/Sdk/DTObjectTemplating/ObjectTemplating/Schema/TemplateSchema.cs]: Copied from SDK

File: src/Sdk/DTObjectTemplating/ObjectTemplating/TemplateEvaluator.cs [hunk 8]
```diff
@@ -47,7 +47,7 @@ internal partial class TemplateEvaluator
             var evaluator = new TemplateEvaluator(context, template, removeBytes);
```
Comment (TingluoHuang) [inline src/Sdk/DTObjectTemplating/ObjectTemplating/TemplateEvaluator.cs]: Copied from SDK to support `required`.

File: src/Sdk/DTObjectTemplating/ObjectTemplating/TemplateReader.cs [hunk 9]
```diff
@@ -178,14 +178,15 @@ private TemplateToken ReadValue(DefinitionInfo definition)
             }
```
Comment (TingluoHuang) [inline src/Sdk/DTObjectTemplating/ObjectTemplating/TemplateReader.cs]: Copied from SDK to support `required`

File: src/Runner.Worker/JobExtension.cs (hunk 2 above)
Comment (TingluoHuang) [reply src/Runner.Worker/JobExtension.cs]: sure

File: src/Runner.Worker/action_yaml.json (hunk 4 above)
Comment (TingluoHuang) [reply src/Runner.Worker/action_yaml.json]: I might keep `strategy` since it contains well-known keys fail-fast, job-index, job-total, and max-parallel.

**Pull request merged**